tfx/components/example_gen/utils.py

```diff
@@ -242,35 +261,108 @@ def _retrieve_latest_span_version(
       - If a matching cannot be found for split pattern provided.
   """
 
+  # Match occurences of pattern '{SPAN*}' and capture span width modifier.
+  span_arg_match = re.findall(SPAN_SPEC_REGEX, split.pattern)
+
+  is_match_span = len(span_arg_match) > 0
+  is_match_date = any(spec in split.pattern for spec in DATE_SPECS)
+  if [is_match_span, is_match_date].count(True) > 1:
+    raise ValueError('Either span spec or date specs must be specified '
+                     'exclusively.')
+
   split_pattern = os.path.join(uri, split.pattern)
 
   split_glob_pattern = split_pattern
   split_regex_pattern = _glob_to_regex(split_pattern)
 
-  latest_span = None
-  latest_version = None
-
-  if SPAN_SPEC not in split.pattern:
+  if not is_match_span and not is_match_date:
     if VERSION_SPEC in split.pattern:
-      raise ValueError('Version spec provided, but Span spec is not present')
-    return latest_span, latest_version
+      raise ValueError('Version spec provided, but Span or Date spec is not '
+                       'present.')
+    return (None, None)
 
-  if split.pattern.count(SPAN_SPEC) != 1:
+  if is_match_span and len(span_arg_match) != 1:
     raise ValueError('Only one %s is allowed in %s' %
                      (SPAN_SPEC, split.pattern))
+  elif is_match_date and not all(split.pattern.count(spec) == 1
+                                 for spec in DATE_SPECS):
+    raise ValueError('Exactly one of each date spec is required in %s' %
+                     split.pattern)
+
+  latest_span_elems = None
+  latest_span_int = None
+  latest_version = None
 
-  split_glob_pattern = split_glob_pattern.replace(SPAN_SPEC, '*')
-  split_regex_pattern = split_regex_pattern.replace(
-      SPAN_SPEC, '(?P<{}>.*)'.format(SPAN_PROPERTY_NAME))
+  if is_match_span:
+    # Check if span spec has any width args. Defaults to greedy matching if 
+    # no width modifiers are present.
+    span_width_regex = '.*'
+    if len(span_arg_match[0]) > 0:
+      # Check for proper formatting of span spec with width modifier.
+      # This spec should have the form '{SPAN:xxx}', where the span regex
+      # captures the sequence ':xxx'.
+      if span_arg_match[0][0] != ':':
+        raise ValueError('Span spec with width modifier is improperly '
+                         'formatted: %s' % split.pattern)
+      try:
+        span_width = int(span_arg_match[0][1:])
```

Comment:
can you try get this just like how you get specs?

e.g. use {SPAN:?<width>} instead of {SPAN(.*?)}?

Reply:
I can try it that way, sure. The main reason I did it this way is so we can just have one matching for both `{SPAN}` and `{SPAN:x}`. If we have the capture group to capture the integer, it adds some case work to the code, which i'm wasn't sure about. For example, to check is_match_span, we need to check if `{SPAN}` is in split pattern or if `{SPAN:(.*)}` regex matches the split pattern. any thoughts on this?

Reply:
updated this. new behavior now uses an or regex. for example, for span the regex is now `{SPAN}|{SPAN:(.+?)}` which matches either exactly `{SPAN}` or `{SPAN:x}` (where `x` is a 1+ length string). also removed the tests for testing malformed specs like `{SPANx}` since they don't seem like relevant cases we need to account for.

Reply:
you just need to get the width count from a string, right?
can we use the way like how you get specs (get yyyy number from string)?

Reply:
fixed.  new logic is to used a named group, optimistically search for this named group, and use default greedy regex when named group does not exist